### PR TITLE
Typing: unification error - warn about missing cmi

### PR DESCRIPTION
**Pull request by hhugo**
Warn when unification failed because cmi could not be found.

```
#cat subdir/m.ml
type a = int
type b = a

#cat a.ml
let (a : M.a) = 2

#cat b.ml
let (b : M.b) = 2

#cat o.ml
let _ = A.a = B.b
```

```
ocamlc -c subdir/m.ml
ocamlc -c -I subdir a.ml
ocamlc -c -I subdir b.ml
ocamlc -c o.ml

File "o.ml", line 1, characters 14-17:
Error: This expression has type M.b but an expression was expected of type
         M.a
       M.b is abstract because no corresponding cmi file was found in path.
       M.a is abstract because no corresponding cmi file was found in path.
```

### CODE REVIEW

File: typing/printtyp.ml [hunk 1]
```diff
@@ -1431,12 +1433,31 @@ let explanation unif t3 t4 ppf =
           fprintf ppf "@,Types for tag `%s are incompatible" l1
       | _ -> ()
       end
+  | Tconstr _, _
+  | _, Tconstr _ ->
+      let missing_def = function
+        | Tconstr (p,_,_) ->
```
Comment (hhugo) [inline typing/printtyp.ml]: Should one search into abbrev_memo ?

**Comment by garrigue** [comment]
This looks like a good idea, but you have to be careful that many people actually use cmi hiding as a feature: this allows to make a type abstract a posteriori, without having to pack.
I'm not sure what would be a good criterion to decide whether this is intentional or not.

As a side remark, hiding the cmi may have strange side-effects, so this is not actively encouraged.

**Comment by trefis** [comment]
> I'm not sure what would be a good criterion to decide whether this is intentional or not.
> 
> As a side remark, hiding the cmi may have strange side-effects, so this is not actively encouraged.

Then perhaps the message should present itself as a warning (on by default)?

**Comment by hhugo** [comment]
It just gives extra information to understand an existing error. One doesn't have to make any assumption on whether this is intentional or not.

File: typing/printtyp.ml [hunk 2]
```diff
@@ -1486,6 +1506,11 @@ let unification_error unif tr txt1 ppf txt2 =
         txt2 (type_expansion t2) t2'
         (trace false "is not compatible with type") tr
         (explanation unif mis);
+      if not (is_Tvar t1 || is_Tvar t2)
```
Comment (hhugo) [inline typing/printtyp.ml]: When explaining the type error in the following, 

```
type 'a t
(fun (x : 'a t as 'a) -> x)
```

`t` is somehow not available in `env`.
I've added condition on `is_Tvar` to workaround this, BUT
- I don't know if it is correct
- if so, I don't know why it is correct

**Comment by dbuenzli** [comment]
> As a side remark, hiding the cmi may have strange side-effects, so this is not actively encouraged.

What do you mean by that ? I do this very often to hide internal modules of my libraries so that they cannot be accessed by consumers of the public API (I don't use pack).

**Comment by garrigue** [comment]
It is still supported.
However, you must be aware that you are not only hiding the implementation of your type, but also its variance, since it is often going to be checked on the fully expanded path, and its type constraints if any. This may have some impact on polymorphism, due to the relaxed value restriction.

```
type 'a t = 'a list  (* hidden *)

val f : unit -> 'a Hidden.t  (* exported *)
```

If Hidden were visible, the result of f would be generalizable, but as it's hidden it won't.

The other problem is that you lose the ability to name that type directly.
I suppose that you support this by re-exporting it.
Again, this works, but you might get errors if working on the hidden version.

In general, I would say this is 100% fine if your types do not have constraints and you don't care about variance. It may get more subtle otherwise, in particular it becomes important that all references to the type go through an exported alias.

hidden/a.mli:

```
type 'a t = 'a list
```

b.ml:

```
type 'a t = 'a A.t
let f () :'a A.t = []
let g () : 'a t = []
```

ocaml b.cmo

```
# B.f();;
- : '_a A.t = <abstr>
# B.g();;
- : 'a B.t = <abstr>
```

**Comment by dbuenzli** [comment]
> The other problem is that you lose the ability to name that type directly.
> I suppose that you support this by re-exporting it.

Yes sure I never make a type visible whose cmi is hidden I actually didn't know this was possible.

**Comment by garrigue** [comment]
A relatively common pattern would be to reexport the same type from several modules, wanting to keep them equal:

b.mli:

```
type 'a t = 'a A.t
```

c.mli:

```
type 'a t = 'a A.t
```

Here, the hidden `A.t` is still observable (as it makes `B.t` and `C.t` equal), but you cannot use it directly. You start having problems if you leave it visible in some exported functions. And you may still get some strange behavior if some part of the type checker are too eager at expanding types (as this is not well exercised).

File: typing/printtyp.ml (hunk 2 above)
Comment (hhugo) [inline typing/printtyp.ml]: @garrigue any comment on this ?

Comment (garrigue) [inline typing/printtyp.ml]: I don't understand your example. It doesn't parse.

File: typing/printtyp.ml [hunk 3]
```diff
@@ -1435,6 +1435,26 @@ let explanation unif t3 t4 ppf =
       end
   | _ -> ()
 
+
+let warn_on_missing_def env ppf t =
+  let is_predef = function
+    | Pident ident' -> List.exists (fun (_,ident) -> ident = ident') Predef.builtin_idents
+    | _ -> false in
+  let missing_def = function
+    | Tconstr (p,_,_) when is_predef p -> None
```
Comment (garrigue) [inline typing/printtyp.ml]: Why do you need a special case for Predef? They are already included in env, so the following case should be sufficient.

File: typing/printtyp.ml [hunk 4]
```diff
@@ -1495,7 +1520,7 @@ let unification_error unif tr txt1 ppf txt2 =
 
 let report_unification_error ppf env ?(unif=true)
     tr txt1 txt2 =
-  wrap_printing_env env (fun () -> unification_error unif tr txt1 ppf txt2)
+  wrap_printing_env env (fun () -> unification_error env unif tr txt1 ppf txt2)
```
Comment (garrigue) [inline typing/printtyp.ml:1518]: Note that the `env` received by `report_unification_error` was introduced to support `-short-paths`.
In some places, this is done in a best-effort way, falling back to the path itself when not available.
If you depend on it for what is defined, one would have to check that it is really accurate.
At least disable your check if `env` is `Env.empty`.

Comment (hhugo) [inline typing/printtyp.ml:1518]: Thanks.

**Pull request merged**